Usability improvements around the picker components

carlobeltrame opened the pull request:
Fixes the following points from #3144, among other usability issues:
- [x] The date pickers for the period date range should react to each other. If I select a start date in next July, the end date picker should automatically show July, and not allow to select dates before the start date. This is also true when creating and editing periods in the camp admin.
- [x] When using English as language, the validation for the end time on new schedule entries is wrong [[video]](https://user-images.githubusercontent.com/7566995/200603868-591270b5-31a9-4ae7-8cc4-ed750fa69f61.mp4)
- [x] Clicking inside the text field of a color or date picker should open the picker
- [x] Color and date picker should not require clicking the okay button. Clicking a value in the picker should be enough to fill in the underlying input (and maybe even close the picker)

Before:
![datepicker-before](https://user-images.githubusercontent.com/7566995/202575063-5ff5bd8f-3326-4135-b425-0f4615b4259c.gif)
After:
![datepicker-after](https://user-images.githubusercontent.com/7566995/202575089-01c6b30e-aa70-408e-be13-6c8663d37f68.gif)


Additionally, I re-wrote all the E-Picker component tests to use vue testing library and user-event, which encourages much better testing [principles](https://testing-library.com/docs/guiding-principles) than vue test-utils. While I was at it, I finally started writing a [frontend testing guide](https://github.com/ecamp/ecamp3/wiki/frontend-testing) in the wiki.

Also, I slightly improved the accessibility of the picker components, by adding an aria-label to the trigger button with which the pickers can be opened. This aria-label was the best way I could target this button in the vue testing library tests, which indicated that our component had accessibility flaws before.

TODO (all done for this PR):
- [x] Re-write the Api-Picker component tests with vue testing library as well
- [x] Fix the still broken locale-specific time parsing of the greaterThan_time, and the tests for it
- [x] Tests for the changed picker behaviour

In a next PR (issue #3210):
- [ ] Write frontend tests for the adjusted component behaviour
  - [ ] CreateCampPeriods: Disallow selecting invalid date combinations
  - [ ] DialogPeriodForm: Disallow selecting invalid date combinations
  - [ ] FormScheduleEntryItem: Disallow selecting invalid date combinations
  - [ ] FormScheduleEntryItem: Disallow selecting invalid time combinations
  - [ ] FormScheduleEntryItem: Allow selecting any time when dates not equal
  - [ ] FormScheduleEntryItem: Only allows selecting start dates in any of the camp's periods
  - [ ] FormScheduleEntryItem: Moves end date along when start date is changed
  - [ ] FormScheduleEntryItem: Moves end time along when start time is changed

carlobeltrame committed:
Replace deprecated slot syntax

carlobeltrame committed:
Fix unusual translation

carlobeltrame committed:
Refactor and simplify pickers to be more user friendly

Improve opening and closing behaviour of date picker
Transfer picker value to input field immediately
Allow pickers to select whether they should auto-close when a value is
selected
For pickers which auto-close, allow re-opening the picker by clicking
inside the text field
Replace the two buttons with just one for closing, now that the value is
auto-transferred to the input field
Allow closing the pickers by clicking outside them (and outside the
input field)
Remove / simplify unused and unnecessarily complex code

carlobeltrame committed:
Allow closing pickers using the escape key

carlobeltrame committed:
Disallow selecting invalid date and time combinations

carlobeltrame committed:
Take the locale into account when validating time inputs

carlobeltrame committed:
Adjust tests to removed buttons

carlobeltrame committed:
Validate value when changed, not when clicking anywhere on the page

carlobeltrame committed:
Also allow shorter versions of manually entered dates in picker

carlobeltrame committed:
Clean up handling of the many values in BasePicker

carlobeltrame committed:
Correct documentation comments

carlobeltrame committed:
Add aria label to picker open buttons

carlobeltrame committed:
Translate format validation error

carlobeltrame committed:
Rewrite date picker specs using testing library

carlobeltrame committed:
Reuse code, follow recommendations of VTL

https://testing-library.com/docs/example-react-intl/#translated-components-testing-stategy

carlobeltrame committed:
Improve aria text wording

carlobeltrame committed:
Rewrite time picker specs using testing library

carlobeltrame committed:
Rewrite color picker tests using testing library

carlobeltrame committed:
Improve accessibility and testability of button components

carlobeltrame committed:
Avoid double submit when clicking save button

carlobeltrame committed:
The reload button should not submit the form field

carlobeltrame committed:
Rewrite api date picker tests using testing library

carlobeltrame committed:
Rewrite api time picker tests using testing library

carlobeltrame committed:
Rewrite api color picker tests using testing library

manuelmeister commented on frontend/src/components/buttons/ButtonEdit.vue:
```diff
@@ -1,7 +1,7 @@
   <v-btn class="px-3 px-sm-4" color="primary" v-bind="$attrs" v-on="$listeners">
     <v-icon :left="$vuetify.breakpoint.smAndUp" size="150%">{{ icon }}</v-icon>
-    <span class="d-none d-sm-block">
+    <span class="d-sr-only d-sm-block">
       <slot>{{ $tc('global.button.edit') }}</slot>
     </span>
```
Why did you do that? For me this hides the text on desktop

carlobeltrame replied:
Hm the goal was to have it always visible for screen readers, and on devices with size small and upwards. Didn't test it though. How would you do it?

manuelmeister replied:
It would need a `d-sm-not-sr-only`. For the moment, we could try using JS breakpoints. Or we could just use aria label and then we don't have the visual problem.

carlobeltrame replied:
Yes, maybe aria-label is easiest here. Will do that for now.

carlobeltrame replied:
Because this component has a slot, it was easier to add a separate (duplicate) instance of `<slot />` which is only displayed on screen readers, and only if the original slot isn't already displayed.

carlobeltrame committed:
Rework accessibility of generic button components

carlobeltrame committed:
Add tests for opening and closing the pickers

carlobeltrame committed:
Add a test for supporting short date notation

carlobeltrame committed:
Extract and unit test our date/time validations

carlobeltrame committed:
Display days of adjacent months in date picker, for easier selection and to make it clear why the picker has this fixed height

carlobeltrame committed:
Initialize veeValidate after dayjs, to make sure dayjs is available for our custom validation rules

manuelmeister reviewed (approved):
Very cool!

There are various visual problems, that still bother me. For example, the many icons of the pickers. They make the UI so cluttered. 

But I like the way this is going much better, so I will try to find a better visual solution once this PR is merged.

carlobeltrame committed:
Update tests after newly displaying adjacent months' days

carlobeltrame committed:
Try to display a useful month when opening the date picker

usu commented on frontend/src/components/program/FormScheduleEntryItem.vue:
```diff
@@ -8,7 +8,7 @@
           vee-id="startDate"
           vee-rules="required"
-          :allowed-dates="allowedStartDates"
+          :max="localScheduleEntry.start"
```
Not sure I understand this logic:
- I can pick a date that is maximum the start date of the scheduleEntry? 
- before with allowedDates, the user was able to select any valid date within a period. With min/max this restriction will not be possible


**Before**
![image](https://user-images.githubusercontent.com/449555/205656658-206a3830-d16a-482d-ad24-978db7f164aa.png)


**Now**
![image](https://user-images.githubusercontent.com/449555/205657162-bc2c2936-bb97-4202-9744-4e7e90d39f8e.png)

carlobeltrame replied:
Uhm, yes, that is a bug, thanks. Should day `:max="localScheduleEntry.end"`. As you said on the other comment, we can improve this if we implement the common behaviour of having the end date move when the start date is moved. But on the end date picker it should still say `:min="localScheduleEntry.start"`. I will implement that and write tests for this component.

And yes, I can additionally restrict the date picker to the periods in this form. I didn't know we support moving schedule entries across periods. As long as we support this, this is another blocker for #2454. I have now added it there.

carlobeltrame replied:
Fixed in [2ebe629](https://github.com/ecamp/ecamp3/pull/3172/commits/2ebe629ab3e8bac8756b26a7393185fb01f6f496)

usu commented on frontend/src/components/form/base/BasePicker.vue:
```diff
@@ -3,11 +3,11 @@ Displays a field as a picker (can be used with v-model)
-      ref="menu"
       v-model="showPicker"
       :disabled="disabled || readonly"
+      :close-on-click="false"
```
Why is this needed? The effect is now, that pickers are not closing when I click on another picker activator.

![image](https://user-images.githubusercontent.com/449555/205658134-9faf8d4f-dec6-4cc6-a72a-3a767165597b.png)

carlobeltrame replied:
It is needed because the timepicker requires multiple clicks to select a time (first click an hour, then click a minute). TBH I have no idea how the overlay stayed open when interacting with the time picker on the old implementation. But I tried to alleviate this by implementing close-on-click-outside and close-on-escape. Unfortunately I cannot listen to activator clicks with `.preventDefault()`, and I can't use the text field's focus state either, because interacting with the picker de-focuses all text fields.

usu replied:
It actually shouldn't because you already have `:close-on-content-click="false"` on the next line. At least in chrome it's working for me, even if I remove `:close-on-click="false"`

carlobeltrame replied:
Thanks, you were right. Fixed in [a407ffd](https://github.com/ecamp/ecamp3/pull/3172/commits/a407ffd92d19357dbc0721e848db6d39561566b3). Only one picker will now be open at any given time.

usu commented on frontend/src/components/program/FormScheduleEntryItem.vue:
```diff
@@ -19,6 +19,7 @@
           :name="$tc('components.program.formScheduleEntryItem.start')"
           vee-id="startDatetime"
           vee-rules="required"
+          :max="maxStartTime"
```
Before, I could shift the start time to any time of the day, which shifted the entry and the duration.

Now this is not possible anymore: To shift an entry to a later date, I have to first change the end time and only then I can change the start time. But then I have to change the end date again, because the duration was kept. 😅 
This seems akward to me.

manuelmeister replied:
I think, in the future we should implement a single range datetime picker. Where you can see the duration and also setup the time as well.

carlobeltrame replied:
+1 for a range datetime picker in the future. For now: Good point, I was not aware of any logic we have so far involving auto-shifting the end date with the start date. I can have a look and re-activate that.

usu replied:
> I was not aware of any logic we have so far involving auto-shifting the end date with the start date. I can have a look and re-activate that.

The log is still there & enabled:
https://github.com/ecamp/ecamp3/pull/3172/files#diff-0e34dccf17184e2244a9d147e7ef554a40281d2d5431375a80bce90ccd099386R149

carlobeltrame replied:
Fixed in [2ebe629](https://github.com/ecamp/ecamp3/pull/3172/commits/2ebe629ab3e8bac8756b26a7393185fb01f6f496)

usu commented on frontend/src/components/form/base/EColorPicker.vue:
```diff
@@ -11,22 +11,18 @@ Displays a field as a color picker (can be used with v-model)
-        </v-btn>
-        <v-btn text color="primary" data-testid="action-ok" @click="picker.on.save">
-          {{ $tc('global.button.ok') }}
+        <v-btn text color="primary" @click="picker.close">
```
If we still need a close button, why not keeping the ok/cancel buttons, so the user has both functionalities? I don't think this is necessarily an improvement. But I can understand, there are different views on this.

carlobeltrame replied:
The pickers as I understand them intuitively (and find twice as efficient to use, see the screencaps in the PR description) is that they are not modal dialogs, but rather input controls, just like a text field or number input. So there is no "old saved value" to go back to with a cancel button. When interacting with the picker, it immediately updates the text field. Interacting with the picker is in every way interchangeable with just editing the value in the text field. This leads to better keyboard accessibility and testability.

This difference, that the pickers are modal dialogs instead of "specialized on-screen keyboards" for entering values into the picker text field, is the thing I find the most annoying by far with the current picker implementation, and it's a big part of the reason why I wanted to make these improvements in the first place.

manuelmeister replied:
I think, once we have a range datetime picker, we can reintroduce the cancel and ok behavior. If needed.

I agree with @carlobeltrame. As I wrote in my comment, the current PR has some minor visual problems. E.g. I'd rather have a floating close "X" outside the panel instead of this "Cancel" button.

usu commented on frontend/src/components/campAdmin/DialogPeriodForm.vue:
```diff
@@ -11,12 +11,14 @@
       :name="$tc('entity.period.fields.start')"
       vee-id="start"
       vee-rules="required"
+      :max="localPeriod.end"
```
This one depends a bit on how we progress in https://github.com/ecamp/ecamp3/pull/3127/
Depending on the edit mode this can/cannot make sense

carlobeltrame replied:
Absolutely. I just wanted to leave the code in a consistent state, no matter what the decision in #3127 comes out to be. I didn't want to introduce client-side validation on period create, but not on update.

carlobeltrame committed:
Reactivate the capability to move schedule entries to different periods via the popup dialog

carlobeltrame committed:
Merge remote-tracking branch 'origin/devel' into picker-fixes

carlobeltrame committed:
Simplify picker closing behaviour

carlobeltrame committed:
Use 12 hour AM/PM format in time picker depending on locale

carlobeltrame committed:
Fix time greater_than validation for different locales

Make the validation more flexible to handle both strings and dayjs
objects as min argument.
Also make tests more complete and clarify edge cases of the dayjs
parser.
